### PR TITLE
Move old ImageDecoder to legacy module and make the nvImageCodec based ImageDecoder the default

### PR DESCRIPTION
**Pull request by jantonguirao**
## Category:

**Other** Replaces default ImageDecoders with the nvImageCodec based implementations. 

## Description:

- Replaces default ImageDecoders with the nvImageCodec based implementations. 
- Move former default decoders to a `legacy` module

### Affected modules and functionalities:

- ImageDecoder*

### Tests:

- [x] Existing tests apply
- [ ] New tests added
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [x] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

File: dali/operators/imgcodec/host_decoder.cc [hunk 1]
```diff
@@ -32,10 +32,26 @@ using HostDecoderCrop = WithCropAttr<HostDecoder, CPUBackend>;
...
+
+// Deprecated aliases: fn.image*_decoder
+DALI_REGISTER_OPERATOR(ImageDecoder, HostDecoder, CPU);
+DALI_REGISTER_OPERATOR(ImageDecoderCrop, HostDecoderCrop, CPU);
+DALI_REGISTER_OPERATOR(ImageDecoderSlice, HostDecoderSlice, CPU);
+DALI_REGISTER_OPERATOR(ImageDecoderRandomCrop, HostDecoderRandomCrop, CPU);
+
+// Deprecated aliases: fn.image*_decoder
 DALI_REGISTER_OPERATOR(experimental__decoders__Image, HostDecoder, CPU);
+DALI_REGISTER_OPERATOR(experimental__ImageDecoder, HostDecoder, CPU);
```
Comment (stiepan) [inline dali/operators/imgcodec/host_decoder.cc:48]: Why add those now if we have not exposed them in this form before? We don't break any existing code by not having them.

image

File: dali/operators/decoder/nvjpeg/nvjpeg_decoder_decoupled_api_test.cc
Comment (stiepan) [inline dali/operators/decoder/nvjpeg/nvjpeg_decoder_decoupled_api_test.cc:1]: Why removing all those tests?

File: dali/operators/imgcodec/mixed_decoder.cc [hunk 2]
```diff
@@ -33,11 +33,26 @@ using MixedDecoderCrop = WithCropAttr<MixedDecoder, MixedBackend>;
...
+
+// Deprecated aliases: fn.image*_decoder
+DALI_REGISTER_OPERATOR(ImageDecoder, MixedDecoder, Mixed);
+DALI_REGISTER_OPERATOR(ImageDecoderCrop, MixedDecoderCrop, Mixed);
+DALI_REGISTER_OPERATOR(ImageDecoderSlice, MixedDecoderSlice, Mixed);
+DALI_REGISTER_OPERATOR(ImageDecoderRandomCrop, MixedDecoderRandomCrop, Mixed);
+
+// Deprecated aliases: fn.experimental.decoders.image*
 DALI_REGISTER_OPERATOR(experimental__decoders__Image, MixedDecoder, Mixed);
+DALI_REGISTER_OPERATOR(experimental__ImageDecoder, MixedDecoder, Mixed);
```
Comment (stiepan) [inline dali/operators/imgcodec/mixed_decoder.cc:49]: Like for the host_decoder, do we need new deprecated name?

File: dali/operators/imgcodec/image_decoder.h [hunk 3]
```diff
@@ -597,8 +615,17 @@ class ImageDecoder : public StatelessOperator<Backend> {
...
     } else {
-      st.image_info.sample_format = NVIMGCODEC_SAMPLEFORMAT_I_RGB;
-      st.orig_img_type = DALI_RGB;
+      // TIFF decoder has some issues with decoding RGB outputs on 1-channel inputs on nvImageCodec
+      // 0.2.0 In that case, decode directly to grayscale and convert later
+      if (!version_at_least(0, 3, 0) &&
+          st.parsed_sample.nvimgcodec_img_info.sample_format == NVIMGCODEC_SAMPLEFORMAT_P_Y) {
+        st.image_info.sample_format = NVIMGCODEC_SAMPLEFORMAT_P_Y;
+        st.orig_img_type = DALI_GRAY;
+        decode_shape[2] = 1;
```
Comment (stiepan) [inline dali/operators/imgcodec/image_decoder.h:624]: nitpick

Wouldn't the direct st.out_shape write be more readable? It looks like we need the `decode_shape` only in this one place in the older version fallback. 

```suggestion
        st.out_shape[2] = 1;
```

File: dali/test/python/checkpointing/test_dali_checkpointing.py [hunk 4]
```diff
@@ -1176,7 +1193,6 @@ def pipeline():
 unsupported_ops = [
     "experimental.decoders.video",
     "experimental.inputs.video",
-    "experimental.decoders.image_random_crop",
```
Comment (stiepan) [inline dali/test/python/checkpointing/test_dali_checkpointing.py:1179]: Thanks!

File: dali/operators/decoder/nvjpeg/nvjpeg_decoder_decoupled_api_test.cc
Comment (jantonguirao) [reply dali/operators/decoder/nvjpeg/nvjpeg_decoder_decoupled_api_test.cc:1]: The new decoder doesn't have means to check which implementation was used (HW, CUDA, etc)

File: dali/operators/imgcodec/image_decoder.h [hunk 5]
```diff
@@ -562,18 +573,24 @@ class ImageDecoder : public StatelessOperator<Backend> {
     return true;
   }
 
+  /**
+   * @brief Checks that nvImageCodec version is at least a given version
+   */
+  bool version_at_least(int req_major, int req_minor, int req_patch) {
```
[the remark this answers is not in the document]
Comment (jantonguirao) [reply dali/operators/imgcodec/image_decoder.h:579]: you are totally right

File: dali/operators/imgcodec/image_decoder.h [hunk 6]
```diff
@@ -577,9 +577,10 @@ class ImageDecoder : public StatelessOperator<Backend> {
    * @brief Checks that nvImageCodec version is at least a given version
    */
   bool version_at_least(int req_major, int req_minor, int req_patch) {
-    int major = -1, minor = -1, patch = -1;
+    int req_majormajor = -1, minor = -1, patch = -1;
     get_nvimgcodec_version(&major, &minor, &patch);
-    return major >= req_major && minor >= req_minor && patch >= req_patch;
+    return MAKE_SEMANTIC_VERSION(major, minor, patch) >=
```
Comment (stiepan) [inline dali/operators/imgcodec/image_decoder.h:582]: I didn't know about this macro, is this something generic? :o

**Review by stiepan — APPROVED** [review]

File: dali/operators/imgcodec/image_decoder.h (hunk 6 above)
Comment (szkarpinski) [reply dali/operators/imgcodec/image_decoder.h:582]: Seems to nvimagecodec's https://github.com/NVIDIA/nvImageCodec/blob/bd192034b05991fb368913fe8afcc09f1711d399/include/nvimgcodec_version.h.in#L32

File: dali/operators/imgcodec/image_decoder.h [hunk 7]
```diff
@@ -562,18 +573,25 @@ class ImageDecoder : public StatelessOperator<Backend> {
     return true;
   }
 
+  /**
+   * @brief Checks that nvImageCodec version is at least a given version
+   */
+  bool version_at_least(int req_major, int req_minor, int req_patch) {
+    int req_majormajor = -1, minor = -1, patch = -1;
```
Comment (szkarpinski) [inline dali/operators/imgcodec/image_decoder.h:580]:
```suggestion
    int major = -1, minor = -1, patch = -1;
```

File: dali/operators/imgcodec/peek_image_shape.cc [hunk 8]
```diff
@@ -115,7 +115,7 @@ void ImgcodecPeekImageShape::RunImpl(Workspace &ws) {
 }
 
 
-DALI_REGISTER_OPERATOR(experimental__PeekImageShape, ImgcodecPeekImageShape, CPU);
```
Comment (szkarpinski) [inline dali/operators/imgcodec/peek_image_shape.cc:118]: We probably want to keep `experimental.peek_image_shape` for backwards compatibility as well, right?

**Comment by szkarpinski** [comment]
I wonder if it makes sense to add an alias to (already deprecated) `fn.image_decoder` in `fn.legacy.image_decoder`. This doesn't improve backward compatibility (if somebody was using `fn.image_decoder` and wants to stick with the old decoder they can as well type `fn.legacy.decoders.image` instead of `fn.legacy.image_decoder`)

**Review by szkarpinski — APPROVED** [review]

**Pull request merged**